source/lac/petsc_matrix_base.cc

```diff
@@ -556,6 +556,8 @@ namespace PETScWrappers
                             PETSC_DEFAULT,
                             &result.petsc_matrix());
           AssertThrow(ierr == 0, ExcPETScError(ierr));
+          ierr = PETScWrappers::destroy_matrix(tmp);
```

Comment:
This looks correct to me, we need to delete the temporary. What is ``MatMatMult`` with ``MAT_INITIAL_MATRIX`` doing to the existing ``result`` matrix? This could also be a leak or not?

Reply:
In the google group we discussed on a little dummy example and I shared my code there:
https://groups.google.com/g/dealii/c/ad6-kfYP2z0
-Running on my machine that looked fine (doing the mmult with a vector 1000 times with a large-enough matrix).

Reply:
@tjhei: The documentation doesn't say: https://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/Mat/MatMatMult.html

But the argument is simply a `Mat *`, not a `Mat *&`, and so the function clearly expects a matrix object that has been created via `MatCreate` or similar before because it can't return a matrix that it creates itself. 

The implementation is here: https://www.mcs.anl.gov/petsc/petsc-current/src/mat/interface/matrix.c.html#MatMatMult and calls line 9261 from where we get to this place: https://www.mcs.anl.gov/petsc/petsc-current/src/mat/interface/matproduct.c.html#MatProductCreate from where I'm losing track of where it goes. But I suspect that it keeps the object and replaces the content.

Reply:
I guess if a test confirms that we are not leaking PETSc objects when repeatedly calling this function, then we are good.